### PR TITLE
New api to list out missed rounds miners

### PR DESCRIPTION
**Pull request by wjrjerome**
# Proposed changes
**This PR is working in progress**
Adding a new API to help devs to debug when miners missed to produce block during their term (their round)

## Types of changes

What types of changes does your code introduce to XDC network?
_Put an `✅` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Regular KTLO or any of the maintaince work. e.g code style
- [ ] CICD Improvement

## Impacted Components
Which part of the codebase this PR will touch base on,

_Put an `✅` in the boxes that apply_

- [ ] Consensus
- [ ] Account
- [ ] Network
- [ ] Geth
- [ ] Smart Contract
- [ ] External components
- [ ] Not sure (Please specify below)
- [x] API

## Checklist
_Put an `✅` in the boxes once you have confirmed below actions (or provide reasons on not doing so) that_

- [x] This PR has sufficient test coverage (unit/integration test) OR I have provided reason in the PR description for not having test coverage
- [ ] Provide an end-to-end test plan in the PR description on how to manually test it on the devnet/testnet.
- [x] Tested the backwards compatibility.
- [x] Tested with XDC nodes running this version co-exist with those running the previous version.
- [ ] Relevant documentation has been updated as part of this PR
- [ ] N/A

### CODE REVIEW

File: consensus/XDPoS/engines/engine_v2/utils.go [hunk 1]
```diff
@@ -163,3 +163,52 @@ func (x *XDPoS_v2) GetSignersFromSnapshot(chain consensus.ChainReader, header *t
...
+			return nil, err
+		}
+		currRound, err := x.GetRoundNumber(nextHeader)
+		if err != nil {
+			return nil, err
+		}
+		// This means there is a missing round incrementation when producing blocks
+		if parentRound+1 != currRound {
+			// We need to loop through between parentRound and the currRound to assess which miner did not mine
+			for i := parentRound; i < currRound; i++ {
```
Comment (liam-lai) [inline consensus/XDPoS/engines/engine_v2/utils.go:190]: i < currRound - 1

Comment (liam-lai) [inline consensus/XDPoS/engines/engine_v2/utils.go:190]: i := parentRound+1

File: consensus/XDPoS/engines/engine_v2/utils.go [hunk 2]
```diff
@@ -163,3 +163,52 @@ func (x *XDPoS_v2) GetSignersFromSnapshot(chain consensus.ChainReader, header *t
...
+				leaderIndex := uint64(i) % x.config.Epoch % uint64(len(masternodes))
+				whosTerm := masternodes[leaderIndex]
+				missedRounds = append(
+					missedRounds,
+					utils.MissedRoundInfo{
+						Round:            i,
+						Miner:            whosTerm,
```
Comment (liam-lai) [inline consensus/XDPoS/engines/engine_v2/utils.go:197]: Turn

Comment (liam-lai) [reply consensus/XDPoS/engines/engine_v2/utils.go:197]: whosTerm.Hex()

File: internal/web3ext/web3ext.go [hunk 3]
```diff
@@ -156,6 +156,11 @@ web3._extend({
 			name: 'getLatestPoolStatus',
 			call: 'XDPoS_getLatestPoolStatus'
 		}),
+		new web3._extend.Method({
+			name: 'GetMissiedRoundsInEpochByBlockHash',
+			call: 'XDPoS_GetMissiedRoundsInEpochByBlockHash',
```
Comment (liam-lai) [inline internal/web3ext/web3ext.go:161]: can you do by BlockNumber?

File: consensus/XDPoS/engines/engine_v2/utils.go [hunk 4]
```diff
@@ -163,3 +163,52 @@ func (x *XDPoS_v2) GetSignersFromSnapshot(chain consensus.ChainReader, header *t
...
+						CurrentBlockHash: nextHeader.Hash(),
+						ParentBlockHash:  parentHeader.Hash(),
```
Comment (liam-lai) [inline consensus/XDPoS/engines/engine_v2/utils.go:199]: I would like block number, not hash

**Comment by liam-lai** [comment]
change pr merge into dev-upgrade

File: consensus/XDPoS/engines/engine_v2/utils.go (hunk 4 above)
Comment (wjrjerome) [reply consensus/XDPoS/engines/engine_v2/utils.go:199]: i give u both

File: internal/web3ext/web3ext.go (hunk 3 above)
Comment (wjrjerome) [reply internal/web3ext/web3ext.go:161]: of course

**Review by liam-lai — APPROVED** [review]

**Pull request merged**